integrations/access/accessrequest/app.go

```diff
@@ -0,0 +1,447 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	reqData := pd.AccessRequestData{
+		User:              req.GetUser(),
+		Roles:             req.GetRoles(),
+		RequestReason:     req.GetRequestReason(),
+		SystemAnnotations: req.GetSystemAnnotations(),
+		Resources:         resourceNames,
+	}
+
+	_, err = a.pluginData.Create(ctx, reqID, PluginData{AccessRequestData: reqData})
+	switch {
+	case err == nil:
+		// This is a new access-request, we have to broadcast it first.
+		if recipients := a.getMessageRecipients(ctx, req); len(recipients) > 0 {
+			if err := a.broadcastAccessRequestMessages(ctx, recipients, reqID, reqData); err != nil {
+				return trace.Wrap(err)
+			}
+		} else {
+			log.Warning("No channel to post")
+		}
+	case trace.IsAlreadyExists(err):
+		// The messages were already sent, nothing to do, we can update the reviews
+	default:
+		// This is an unexpected error, returning
+		return trace.Wrap(err)
+	}
+
+	// This is an already existing access request, we post reviews and update its status
+	if reqReviews := req.GetReviews(); len(reqReviews) > 0 {
+		if err := a.postReviewReplies(ctx, reqID, reqReviews); err != nil {
+			return trace.Wrap(err)
+		}
+
+		err := a.updateMessages(ctx, reqID, pd.Unresolved, "", reqReviews)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	// We always post review replies in thread. If the messaging service does not support
+	// threading this will do nothing
+	replyErr := a.postReviewReplies(ctx, req.GetName(), req.GetReviews())
+
+	reason := req.GetResolveReason()
+	state := req.GetState()
+	var tag pd.ResolutionTag
+
+	switch state {
+	case types.RequestState_APPROVED:
+		tag = pd.ResolvedApproved
+	case types.RequestState_DENIED:
+		tag = pd.ResolvedDenied
+	case types.RequestState_PROMOTED:
+		tag = pd.ResolvedPromoted
+	default:
+		logger.Get(ctx).Warningf("Unknown state %v (%s)", state, state.String())
+		return replyErr
+	}
+	err := trace.Wrap(a.updateMessages(ctx, req.GetName(), tag, reason, req.GetReviews()))
+	return trace.NewAggregate(replyErr, err)
+}
+
+func (a *App) onDeletedRequest(ctx context.Context, reqID string) error {
+	return a.updateMessages(ctx, reqID, pd.ResolvedExpired, "", nil)
+}
+
+// broadcastAccessRequestMessages sends nessages to each recipient for an access-request.
+// This method is only called when for new access-requests.
+func (a *App) broadcastAccessRequestMessages(ctx context.Context, recipients []common.Recipient, reqID string, reqData pd.AccessRequestData) error {
+	sentMessages, err := a.bot.BroadcastAccessRequestMessage(ctx, recipients, reqID, reqData)
+	if len(sentMessages) == 0 && err != nil {
+		return trace.Wrap(err)
+	}
+	for _, data := range sentMessages {
+		logger.Get(ctx).WithFields(logger.Fields{
+			"channel_id": data.ChannelID,
+			"message_id": data.MessageID,
+		}).Info("Successfully posted messages")
+	}
+	if err != nil {
```

Comment:
Is it even possible for this to be true? Can you have a non-empty sentMessages slice and an error?

Reply:
It'll be highly dependent on the Bot implementation on whether this is possible.